TrafficReplayer/src/org/opensearch/migrations/replay/TrafficReplayer.java

```diff
@@ -78,26 +87,49 @@ public static void main(String[] args) throws IOException, InterruptedException
             exitWithUsageAndCode(3);
             return;
         }
+        Path outputPath;
+        try {
+            outputPath = Paths.get(args[2]);
+        } catch (Exception e) {
+            exitWithUsageAndCode(4);
+            return;
+        }
+
         var tr = new TrafficReplayer(uri);
-        ReplayEngine replayEngine = new ReplayEngine(rp->tr.writeToSocketAndClose(rp));
-        tr.runReplay(directoryPath, replayEngine);
+        try (var tripleWriter = new TripleToFileWriter(outputPath)){
+            ReplayEngine replayEngine = new ReplayEngine(
+                    rp -> tr.writeToSocketAndClose(rp, (triple) -> tripleWriter.write(triple))
+            );
+            tr.runReplay(directoryPath, replayEngine);
+        }
     }
 
     private static String aggregateByteStreamToString(Stream<byte[]> bStream) {
         return new String(Bytes.concat(bStream.toArray(byte[][]::new)), Charset.defaultCharset());
     }
 
     static int nReceived = 0;
-    private void writeToSocketAndClose(RequestResponsePacketPair requestResponsePacketPair) {
+    private void writeToSocketAndClose(RequestResponsePacketPair requestResponsePacketPair,
+                                       Consumer<RequestResponseResponseJSONTriple> onResponseReceivedCallback) {
         try {
             log("Assembled request/response - starting to write to socket");
             var packetHandler = packetHandlerFactory.create();
             for (var packetData : requestResponsePacketPair.requestData) {
                 packetHandler.consumeBytes(packetData);
             }
             AtomicBoolean waiter = new AtomicBoolean(true);
-            packetHandler.finalizeRequest(summary->{
+            var requestResponseTriple = new RequestResponseResponseJSONTriple(requestResponsePacketPair);
+            packetHandler.finalizeRequest(summary-> {
                 System.err.println("Summary(#"+(nReceived++)+")="+summary);
+                try {
+                    requestResponseTriple.addShadowResponse(summary.getReceiptTimeAndResponsePackets().map(entry -> entry.getValue()).collect(Collectors.toList()),
+                            summary.getResponseDuration());
+                } catch (IOException e) {
+                    e.printStackTrace();
```

Comment:
You should add a very noisy log statement here too that also says something like - SWALLOWING EXCEPTION, XXXXXX, TODO, etc.  TODO in the code is helpful too.
That's basic programming tip... if I have a byte array, what could possibly go wrong?  Why would the block above ever NEED to throw an IOException.  If it's an internal error that should never happen, that should be handled further down the stack - this code shouldn't be concerned with it & likely doesn't have the means to deal with it, since any exceptions wouldn't be something that this code has any control over.